### PR TITLE
Got target node name of pending DaemonSet pod.

### LINKED ISSUE #61050: [test failed] gci-gce-alpha-features
https://k8s-testgrid.appspot.com/sig-release-master-blocking#gci-gce-alpha-features
testgrid diff for commit range: https://github.com/kubernetes/kubernetes/compare/f7d9fb474...f7aafaeb4

looks like #59862 

/priority failing-test
/priority critical-urgent
/kind bug
/status approved-for-milestone
/sig apps
/milestone v1.10

cc @jdumars @jberkus 

/assign @k82cn @janetkuo

### PR DESCRIPTION
**Pull request by k82cn**
Signed-off-by: Da K. Ma <klaus1982.cn@gmail.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
part of #59194 Fixes #61050

**Release note**:
```release-note
None
```

### CODE REVIEW

**Comment by k82cn** [comment]
There's a race condition that: after DaemonSet controller created the pods, scheduler takes time to bind the pod to the node; if node condition was changed (e.g. label updated) during this period, a new pod was created unexpected. Refer to the following log, thousands of pods was created :(

```
I0320 01:25:18.592] Mar 20 01:25:18.385: INFO: 24 / 2723 pods in namespace 'kube-system' are running and ready (608 seconds elapsed)
```

**Comment by liggitt** [comment]
> There's a race condition that: after DaemonSet controller created the pods, scheduler takes time to bind the pod to the node; if node condition was changed (e.g. label updated) during this period, a new pod was created unexpected. Refer to the following log, thousands of pods was created

Why is the daemonset creating pods that are sensitive to node labels?

File: pkg/controller/daemon/util/daemonset_util.go [hunk 1]
```diff
@@ -225,3 +225,43 @@ func AppendNoScheduleTolerationIfNotExist(tolerations []v1.Toleration) []v1.Tole
...
+
+		terms := pod.Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms
+		if len(terms) < 1 {
+			return "", fmt.Errorf("no nodeSelectorTerms in requiredDuringSchedulingIgnoredDuringExecution of pod %s/%s",
+				pod.Namespace, pod.Name)
+		}
+
+		for _, term := range terms {
+			for _, exp := range term.MatchExpressions {
+				if exp.Key == kubeletapis.LabelHostname &&
```
Comment (liggitt) [inline pkg/controller/daemon/util/daemonset_util.go]: it appears the daemonset depends on the value of this label matching the node name. that is *not* a valid assumption.

Comment (liggitt) [reply pkg/controller/daemon/util/daemonset_util.go]: opened https://github.com/kubernetes/kubernetes/issues/61410 to track

**Comment by k82cn** [comment]
>> There's a race condition that: after DaemonSet controller created the pods, scheduler takes time to bind the pod to the node; if node condition was changed (e.g. label updated) during this period, a new pod was created unexpected. Refer to the following log, thousands of pods was created

> Why is the daemonset creating pods that are sensitive to node label changes?

DaemonSet will create pods to the selected nodes, one of selection is node selector for node labels; so if label of node changed, DaemonSet controller should delete or create pods accordingly.

**Comment by k82cn** [comment]
>> There's a race condition that: after DaemonSet controller created the pods, scheduler takes time to bind the pod to the node; if node condition was changed (e.g. label updated) during this period, a new pod was created unexpected. Refer to the following log, thousands of pods was created

> Why is the daemonset creating pods that are sensitive to node label changes?

DaemonSet will create pods to the selected nodes, one of selection is node selector for node labels; so if label of node changed, DaemonSet controller should delete or create pods accordingly.

**Comment by k82cn** [comment]
Thanks for liggitt@ 's comments, I'll continue this PR after https://github.com/kubernetes/kubernetes/issues/61410 merged :)

**Comment by k82cn** [comment]
Handled this in #63223 :)

**Pull request closed without merging**